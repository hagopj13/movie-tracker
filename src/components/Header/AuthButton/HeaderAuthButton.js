// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';

import dialogTypes from 'config/dialogTypes';
import authSelectors from 'store/auth/auth.selectors';
import authActions from 'store/auth/auth.actions';
import AuthActionTypes from 'store/auth/auth.types';
import loadingSelectors from 'store/api/loading/loading.selectors';
import dialogActions from 'store/ui/dialog/dialog.actions';

type Props = {
  isAuth: boolean,
  isLogoutLoading: boolean,
  onLogout: () => void,
  onShowLoginDialog: () => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    minWidth: 72,
    textTransform: 'none',
    fontSize: theme.typography.h6.fontSize,
    '&$disabled': {
      color: theme.palette.grey[300],
    },
  },
  disabled: {},
  logoutText: {
    color: theme.palette.common.white,
  },
}));

export const HeaderAuthButton = (props: Props) => {
  const { isAuth, isLogoutLoading, onLogout, onShowLoginDialog } = props;
  const classes = useStyles();

  return isAuth ? (
    <Button
      classes={{ root: classes.root, disabled: classes.disabled, text: classes.logoutText }}
      disabled={isLogoutLoading}
      onClick={onLogout}
    >
      Logout
    </Button>
  ) : (
    <Button className={classes.root} color="primary" onClick={onShowLoginDialog}>
      Login
    </Button>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: authSelectors.selectIsAuth,
  isLogoutLoading: loadingSelectors.createIsLoadingSelector([AuthActionTypes.LOGOUT]),
});

const mapDispatchToProps = {
  onLogout: authActions.logout,
  onShowLoginDialog: () => dialogActions.showDialog(dialogTypes.LOGIN),
};

export default connect(mapStateToProps, mapDispatchToProps)(HeaderAuthButton);
