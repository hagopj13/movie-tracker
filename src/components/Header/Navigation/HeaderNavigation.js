// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { NavLink } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Link from '@material-ui/core/Link';

import authSelectors from 'store/auth/auth.selectors';
import navigationItems from 'config/navigationItems';

type Props = {
  isAuth: boolean,
};

const useStyles = makeStyles((theme) => ({
  navLink: {
    color: theme.palette.common.white,
    marginRight: theme.spacing(3),
    '&:last-child': {
      marginRight: 0,
    },
  },
  selectedNavLink: {
    color: theme.palette.primary.main,
  },
}));

export const HeaderNavigation = (props: Props) => {
  const { isAuth } = props;

  const classes = useStyles();

  return (
    <div>
      {navigationItems.map(
        (navItem) =>
          (isAuth || !navItem.requiresAuth) && (
            <Link
              key={navItem.id}
              component={NavLink}
              exact
              to={navItem.to}
              className={classes.navLink}
              activeClassName={classes.selectedNavLink}
            >
              <Typography variant="h6" component="span">
                {navItem.label}
              </Typography>
            </Link>
          ),
      )}
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: authSelectors.selectIsAuth,
});

export default connect(mapStateToProps)(HeaderNavigation);
