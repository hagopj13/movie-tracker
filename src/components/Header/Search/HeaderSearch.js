// @flow
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { useHistory, useLocation } from 'react-router-dom';

import searchActions from 'store/search/search.actions';
import searchSelectors from 'store/search/search.selectors';

import SearchForm from 'components/Search/Form/SearchForm';

type Props = {
  selectedSearchQuery: string,
  onSetSearchQuery: (query: string) => void,
};

export const HeaderSearch = (props: Props) => {
  const { selectedSearchQuery, onSetSearchQuery } = props;

  const history = useHistory();
  const location = useLocation();

  useEffect(() => {
    if (location.pathname === '/search') {
      const urlSearchParams = new URLSearchParams(location.search);
      const urlSearchQuery = urlSearchParams.get('query');
      if (urlSearchQuery && urlSearchQuery !== selectedSearchQuery) {
        onSetSearchQuery(urlSearchQuery);
      }
    }
  }, [location, selectedSearchQuery, onSetSearchQuery]);

  const handleSubmit = ({ query }) => {
    if (query) {
      history.push(`/search?query=${query}`);
    }
  };

  return <SearchForm onSubmit={handleSubmit} selectedQuery={selectedSearchQuery} />;
};

const mapStateToProps = createStructuredSelector({
  selectedSearchQuery: searchSelectors.selectQuery,
});

const mapDispatchToProps = {
  onSetSearchQuery: searchActions.setSearchQuery,
};

export default connect(mapStateToProps, mapDispatchToProps)(HeaderSearch);
