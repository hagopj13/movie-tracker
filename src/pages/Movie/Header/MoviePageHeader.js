// @flow
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import Card from '@material-ui/core/Card';

import MovieBackdrop from 'components/Movie/Backdrop/MovieBackdrop';
import MoviePoster from 'components/Movie/Poster/MoviePoster';
import type { MovieDetails } from 'types';

type Props = {
  movie: MovieDetails,
};

const useStyles = makeStyles((theme) => ({
  root: {
    color: theme.palette.common.white,
    padding: theme.spacing(5),
  },
  poster: {
    borderRadius: 8,
    width: 300,
  },
}));

const MoviePageHeader = (props: Props) => {
  const { movie } = props;

  const classes = useStyles();

  return (
    <MovieBackdrop backdropPath={movie.backdropPath}>
      <Container className={classes.root}>
        <Card className={classes.poster} elevation={5}>
          <MoviePoster posterPath={movie.posterPath} height={450} width={300} />
        </Card>
      </Container>
    </MovieBackdrop>
  );
};

export default MoviePageHeader;
