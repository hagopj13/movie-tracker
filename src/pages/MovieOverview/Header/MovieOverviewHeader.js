// @flow
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import Card from '@material-ui/core/Card';

import MovieBackdrop from 'components/Movie/Backdrop/MovieBackdrop';
import MoviePoster from 'components/Movie/Poster/MoviePoster';
import MovieSummary from 'components/Movie/Summary/MovieSummary';
import MovieActions from 'components/Movie/Actions/MovieActions';
import type { Movie } from 'types';

type Props = {
  movie: Movie,
};

const useStyles = makeStyles((theme) => ({
  root: {
    color: theme.palette.common.white,
    padding: theme.spacing(5),
    display: 'flex',
    [theme.breakpoints.down('xs')]: {
      flexDirection: 'column',
      alignItems: 'center',
    },
  },
  poster: {
    borderRadius: 8,
    width: 300,
    minWidth: 300,
    height: 450,
    marginRight: theme.spacing(5),
    [theme.breakpoints.down('xs')]: {
      marginRight: 0,
      marginBottom: theme.spacing(2),
    },
  },
  infoContainer: {
    paddingTop: theme.spacing(1),
  },
  actionsContainer: {
    marginTop: theme.spacing(2),
    [theme.breakpoints.down('xs')]: {
      textAlign: 'center',
    },
  },
}));

const MovieOverviewHeader = (props: Props) => {
  const { movie } = props;

  const classes = useStyles();

  return (
    <MovieBackdrop backdropPath={movie.backdropPath}>
      <Container className={classes.root}>
        <Card className={classes.poster} elevation={5}>
          <MoviePoster posterPath={movie.posterPath} height={450} width={300} />
        </Card>
        <div className={classes.infoContainer}>
          <MovieSummary movie={movie} />
          <div className={classes.actionsContainer}>
            <MovieActions movie={movie} />
          </div>
        </div>
      </Container>
    </MovieBackdrop>
  );
};

export default MovieOverviewHeader;
