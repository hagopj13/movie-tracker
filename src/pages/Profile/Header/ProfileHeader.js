// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import Typography from '@material-ui/core/Typography';
import Avatar from '@material-ui/core/Avatar';
import { blueGrey } from '@material-ui/core/colors';

import profileSelectors from 'store/user/profile/profile.selectors';

type Props = {
  userName: string,
};

const useStyles = makeStyles((theme) => ({
  root: {
    background: `linear-gradient(${blueGrey[700]}, ${blueGrey[600]} 70%, ${blueGrey[500]})`,
  },
  container: {
    padding: theme.spacing(5),
    display: 'flex',
    alignItems: 'center',
    color: theme.palette.common.white,
    [theme.breakpoints.down('xs')]: {
      flexDirection: 'column',
    },
  },
  avatar: {
    fontSize: 40,
    width: theme.spacing(11),
    height: theme.spacing(11),
    backgroundColor: theme.palette.primary.main,
    textTransform: 'uppercase',
    marginRight: theme.spacing(4),
    [theme.breakpoints.down('xs')]: {
      marginRight: 0,
      marginBottom: theme.spacing(2),
    },
  },
}));

export const ProfileHeader = (props: Props) => {
  const { userName } = props;

  const classes = useStyles();

  return (
    <div className={classes.root}>
      <Container className={classes.container}>
        <Avatar className={classes.avatar} color="primary">
          {userName?.length > 0 ? userName.charAt(0) : ''}
        </Avatar>
        <Typography className={classes.userName} variant="h4">
          {userName}
        </Typography>
      </Container>
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  userName: profileSelectors.selectUserName,
});

export default connect(mapStateToProps)(ProfileHeader);
