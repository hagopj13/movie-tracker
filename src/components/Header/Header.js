// @flow
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Container from '@material-ui/core/Container';
import IconButton from '@material-ui/core/IconButton';
import Box from '@material-ui/core/Box';
import Hidden from '@material-ui/core/Hidden';
import MenuIcon from '@material-ui/icons/Menu';

import Navigation from './Navigation/HeaderNavigation';
import Search from './Search/HeaderSearch';
import AuthButton from './AuthButton/HeaderAuthButton';

type Props = {
  onOpenDrawer: () => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    backgroundColor: theme.palette.secondary.main,
    color: theme.palette.common.white,
  },
  toolbar: {
    justifyContent: 'center',
    minHeight: 64,
    padding: 0,
  },
  container: {
    display: 'flex',
    alignItems: 'center',
    padding: theme.spacing(0, 4, 0, 5),
    [theme.breakpoints.down('sm')]: {
      paddingLeft: theme.spacing(3.5),
    },
    '& > *': {
      marginRight: theme.spacing(2.5),
      '&:last-child': {
        marginRight: 0,
      },
    },
  },
  offsetToolbar: {
    minHeight: 64,
  },
  menuIcon: {
    color: theme.palette.common.white,
    fontSize: 24,
  },
  searchContainer: {
    [theme.breakpoints.down('xs')]: {
      flexGrow: 1,
    },
  },
}));

const Header = (props: Props) => {
  const { onOpenDrawer } = props;
  const classes = useStyles();

  return (
    <>
      <AppBar className={classes.root} elevation={4}>
        <Toolbar className={classes.toolbar}>
          <Container className={classes.container}>
            <Hidden mdUp>
              <IconButton onClick={onOpenDrawer}>
                <MenuIcon className={classes.menuIcon} />
              </IconButton>
            </Hidden>
            <Hidden smDown>
              <Navigation />
            </Hidden>
            <Hidden xsDown>
              <Box flexGrow={1} />
            </Hidden>
            <Box className={classes.searchContainer}>
              <Search />
            </Box>
            <Hidden smDown>
              <AuthButton />
            </Hidden>
          </Container>
        </Toolbar>
      </AppBar>
      <Toolbar className={classes.offsetToolbar} />
    </>
  );
};

export default Header;
