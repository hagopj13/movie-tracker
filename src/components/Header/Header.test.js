import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import IconButton from '@material-ui/core/IconButton';

import Header from './Header';
import HeaderNavigation from './Navigation/HeaderNavigation';
import HeaderSearch from './Search/HeaderSearch';
import HeaderAuthButton from './AuthButton/HeaderAuthButton';

describe('Header component', () => {
  let shallow;
  const title = 'Some title';
  let mockOpenDrawer;
  let wrapper;

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    mockOpenDrawer = jest.fn();
    const mockProps = {
      title,
      onOpenDrawer: mockOpenDrawer,
    };
    wrapper = shallow(<Header {...mockProps} />);
  });

  it('should render the Header component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the HeaderNavigation component', () => {
    expect(wrapper.exists(HeaderNavigation)).toBe(true);
  });

  it('should render the HeaderSearch component', () => {
    expect(wrapper.exists(HeaderSearch)).toBe(true);
  });

  it('should render the HeaderAuthButton component', () => {
    expect(wrapper.exists(HeaderAuthButton)).toBe(true);
  });

  it('should call the onOpenDrawer prop when icon button is clicked', () => {
    wrapper.find(IconButton).simulate('click');
    expect(mockOpenDrawer).toHaveBeenCalled();
  });
});
