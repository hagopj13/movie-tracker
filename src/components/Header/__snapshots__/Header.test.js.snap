// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render the Header component correctly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(AppBar))
    className="makeStyles-root-1"
    elevation={4}
  >
    <WithStyles(ForwardRef(Toolbar))
      className="makeStyles-toolbar-2"
    >
      <WithStyles(ForwardRef(Container))
        className="makeStyles-container-3"
      >
        <Hidden
          mdUp={true}
        >
          <WithStyles(ForwardRef(IconButton))
            onClick={[MockFunction]}
          >
            <MenuIcon
              className="makeStyles-menuIcon-5"
            />
          </WithStyles(ForwardRef(IconButton))>
        </Hidden>
        <Hidden
          smDown={true}
        >
          <Connect(HeaderNavigation) />
        </Hidden>
        <Hidden
          xsDown={true}
        >
          <Styled(MuiBox)
            flexGrow={1}
          />
        </Hidden>
        <Styled(MuiBox)
          className="makeStyles-searchContainer-6"
        >
          <Connect(HeaderSearch) />
        </Styled(MuiBox)>
        <Hidden
          smDown={true}
        >
          <Connect(HeaderAuthButton) />
        </Hidden>
      </WithStyles(ForwardRef(Container))>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
  <WithStyles(ForwardRef(Toolbar))
    className="makeStyles-offsetToolbar-4"
  />
</Fragment>
`;
